sys/include/chunked_ringbuffer.h

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright (C) 2021 ML!PA Consulting GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    sys_chunk_buffer    chunked Ringbuffer
+ * @ingroup     sys
+ * @brief       Implementation of a Ringbuffer to store chunks of data
+ * @{
+ *
+ * @file
+ * @brief   Chunked Ringbuffer
+ *
+ * A chunked ringbuffer is a ringbuffer that holds chunks of data.
```

Comment:
Suggestion:

> A chunked ringbuffer is a ring buffer that holds a limited number of chunks containing (possibly non-contiguous) data.
>
> Data in chunks can be added incrementally without needing to know their length in advance; any overflow invalidates the chunk being added, and needs to be signaled out of band.
>
> This data structure can be used across threads without need for synchronization.
>
> This implementation is designed to make adding to the buffer fast.

(This is in parts to quickly verify whether I'm reading all this right, but also to allow later users to quickly assess whether this is what they need).

By the way, "without need for synchronization": How, with neither irq-off nor atomics? (I get the impression that chunk_cur is used with atomic access intention, shouldn't that be declared atomic?).

Reply:
The assumption is that the writer is in an ISR and can't be interrupted - will add a clarification.

The reader will disable interrupts when it advances `chunk_cur`, unless I missed a case I think that's all that is needed.

Reply:
TBH I'm way more familiar with atomic semantics than the consequences of ISR disabling -- but I'll review for that again later.

What does strike me now already is that irqs are only touched in crb_consume_chunk. If crb_chunk_foreach is supposed to be peeking (and does not consume the chunk), that should be documented.

Generally, peeking seems not to be guarded. If my understanding is correct, the compiler or machine (say, optimizing a task that spins trying to peek) might optimize away that read if there is nothing at all indicating external change.

Reply:
> Generally, peeking seems not to be guarded.

There is no need to, as peeking only happens on complete chunks.
Chunks that are still being written are not inside the valid chunk array

Reply:
While peeking it is not required, but to determine *that* a chunk is available for reading, some pointer (that is written to from an interrupt context) is read. Thus, a compiler seeing `while !is_peekable) {}` that sees no atomic access or interrupt-off line might inline the peekability check and determine that this won't ever be peekable once found not to be.

Reply:
Worse yet, a compiler (or CPU, but our CPUs don't do that) might reorder the actual buffer reads to happen before the index read that says whether data is here. The read in `_get_complete_chunk` should thus be at least be atomically with Acquire semantics (and the write with Release, though that's probably implied by running in the interrupt). `atomic_load_u8` would do that.

Reply:
54e7db8771aeb5172f19b460b6b74c7ab3d84dda should address this